docs/proposals/20230718-inplace-workload-vertical-scaling.md

```diff
@@ -0,0 +1,813 @@
+                return false, nil
+        }
+
++       if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && isPodResizeInProgress(pod, &apiPodStatus) {
++               // While resize is in progress, periodically call PLEG to update pod cache
++               runningPod := kubecontainer.ConvertPodStatusToRunningPod(kl.getRuntime().Type(), podStatus)
++               if err := kl.pleg.UpdateCache(&runningPod, pod.UID); err != nil {
++                       klog.ErrorS(err, "Failed to update pod cache", "pod", klog.KObj(pod))
++                       return false, err
++               }
++       }
++
+        return false, nil
+ }
+```
+
+The corresponding changes to validation are not expanded here.
+
+### Goals
+
+- Users only need to modify template in workload to achieve vertical scaling for all managed Pods
+- Simple and clear configuration strategy, support grayscale capabilities
+- Provides the rollback function, and attempts to return to the previous state when the capacity expansion and shrinkage fails. If the rollback still fails, no further action is required
+
+### Non-Goals/Future Work
+
+- Implement preemption ensures that all Pods can successfully expand and shrink
+
+## Proposal
+
+### User Stories
+
+Users can start vertical scaling of all managed Pods by modifying the resource field in workload.
+
+### Implementation Details/Notes/Constraints
+
+#### Design Objectives
+
+- Use vscale as a standalone operation rather than integrated in update. For example, for cloneset, three operations will be improved to scale, vscale, and update.
+- Adding the vscaleStrategy field to workload requiring vscale operations allows users to configure strategies to handle different scenarios.
+
+#### CloneSet
+
+##### Extend Definition
+
+We want users to write in PodSpec whether or not each container needs to be restarted.
+
+```go
+// CloneSetVScaleStrategy defines strategies for pods vscale.
+type CloneSetVScaleStrategy struct {
+	MaxProposed *intstr.IntOrString `json:"maxProposed,omitempty"`
+	Paused   bool  `json:"paused,omitempty"`
+}
+```
+
+##### Enhance Expectation
+
+Create a new VScaleExpectations for use by different controllers.
+
+```go
+// VScaleExpectations is an interface that allows users to set and wait on expectations of pods scale.
+type VScaleExpectations interface {
+	ExpectVScale(controllerKey string, action ScaleAction, name string)
+	ObserveVScale(controllerKey string, action ScaleAction, name string)
+	SatisfiedExpectations(controllerKey string) (bool, time.Duration, map[ScaleAction][]string)
+	DeleteExpectations(controllerKey string)
+	GetExpectations(controllerKey string) map[ScaleAction]sets.String
+}
+```
+
+##### Enhance calculateDiffsWithExpectation
+
+Vscale-related parameters need to be added to expectationDiffs
+
+```go
+type expectationDiffs struct {
+	...
+	// vscale identifies whether resources have changed
+	vscale bool
```

Comment:
```go
numVScale int32
```

Reply:
Yes, I should change the quantity here. This design is not perfect for the time being, and we need to consider the vertical scaling of some Pods (not all) in the future. Then I'll update it with a commit.